### ISSUE TITLE
Is is possible to perform unstructured sparsification, and then apply W8A8 quantization?

### ISSUE THREAD

**Issue by zjnyly** [vllm-project/llm-compressor#1051]
I’m wondering if it’s possible to use sparseGPT for sparsification, followed by SmoothQuant and GPTQ for one-shot quantization. Will the quantization stage alter the sparsity configuration set during sparsification? How can I preserve the sparsified weights? Should I apply ConstantPruning before running SmoothQuantModifier and GPTQModifier? Is there a recommended recipe for this process? I would greatly appreciate any guidance. Thank you!

**Comment by zjnyly** [comment]
I noticed the example in [https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_w8a8_int8/README.md](url). But I don't know how and where to insert the `SparseGPTModifier` as well as the `ConstantPruningModifier` in the right order. 
 ```
recipe = [
    SmoothQuantModifier(smoothing_strength=0.8),
    GPTQModifier(targets="Linear", scheme="W8A8", ignore=["lm_head"]),
]
```

**Comment by kylesayrs** [comment]
Hi @zjnyly!

Your recipe request is most similar to our example [2of4_w4a16_recipe.yaml](https://github.com/vllm-project/llm-compressor/blob/5a3f7e32fcb72b1b2411009e8febe33f8528f35c/tests/e2e/vLLM/recipes/WNA16_2of4/2of4_w4a16_recipe.yaml#L4) or our [additional_sparsity_with_quant.yaml](https://github.com/vllm-project/llm-compressor/blob/5a3f7e32fcb72b1b2411009e8febe33f8528f35c/tests/llmcompressor/transformers/obcq/recipes/additional_sparsity_with_quant.yaml#L11) example. As a rule of thumb, sparsity should always be applied before quantization.

Here's an example recipe that hasn't been tested but most closely aligns with what you might be looking for.
```python3
recipe = [
   SparseGPTModifier(sparsity=0.7, percdamp=0.01, mask_structure="0:0", ignore=["lm_head"]),
   SmoothQuantModifier(smoothing_strength=0.8),
   GPTQModifier(targets="Linear", scheme="W8A8", ignore=["lm_head"]),
]
```

Let me know how it goes!

**Comment by kylesayrs** [comment]
The `ConstantPruningModifier` is typically used to preserve sparsity between sparsification and quantization
ex. SparseGPTModifier -> ConstantPruningModifier -> QuantizationModifier

While adding the ConstantPruningModifier after SparseGPTModifier will guarantee that future modifications do not reduce sparsity, it is only necessary for the QuantizationModifier and not SmoothQuantModifier or GPTQModifier, as these modifiers preserve sparsity by default. Understandably, this flow is a little confusing and requires some revisiting.

**Comment by zjnyly** [comment]
> I noticed the example in [https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_w8a8_int8/README.md](url). But I don't know how and where to insert the `SparseGPTModifier` as well as the `ConstantPruningModifier` in the right order.
> 
> ```
> recipe = [
>    SmoothQuantModifier(smoothing_strength=0.8),
>    GPTQModifier(targets="Linear", scheme="W8A8", ignore=["lm_head"]),
> ]
> ```

Thank you for your script! I haven’t downloaded the GSM8K dataset yet to evaluate the accuracy. However, I noticed that the models with 50% and 60% unstructured sparsity produce identical results, while the sequences generated by the model with 70% sparsity are less promising. I believe further fine-tuning with sparsity is necessary.

Additionally, I’m curious whether the KV Cache can be quantized to the int8 format instead of fp8. This functionality is particularly important for me. 

I use this script for quantization
```
DEFAULT_stage:
  DEFAULT_modifiers:
    # SparseGPTModifier:
    #   sparsity: 0.6
    #   mask_structure: 0:0
    #   sequential_update: true
    #   targets: ['re:model.layers.\d*$']
    # SmoothQuantModifier: {smoothing_strength: 0.5}
    QuantizationModifier:
            ignore: ["lm_head"]
            config_groups:
                group_0:
                    weights:
                        num_bits: 8
                        type: int
                        strategy: channel
                        dynamic: false
                        symmetric: true
                    input_activations:
                        num_bits: 8
                        type: int
                        strategy: token
                        dynamic: true
                        symmetric: true
                    targets: ["Linear"]
            kv_cache_scheme:
                num_bits: 8
                type: int
                strategy: tensor
                dynamic: false
                symmetric: true
    ConstantPruningModifier:
      start: 0.0
      targets: ['re:.*q_proj.weight', 're:.*k_proj.weight', 're:.*v_proj.weight', 're:.*o_proj.weight',
        're:.*gate_proj.weight', 're:.*up_proj.weight', 're:.*down_proj.weight']
``` 
And to verify KV Cache is quantized, I loaded the quantized model using `AutoModelForCausalLM`, and printed the KV Cache in 'modeling_llama.py' `LlamaSdpaAttention`, but it is still bfloat16. How to verify if KV Cache is quantized to int8? Thanks!

**Comment by kylesayrs** [comment]
> I haven’t downloaded the GSM8K dataset yet to evaluate the accuracy. However, I noticed that the models with 50% and 60% unstructured sparsity produce identical results, while the sequences generated by the model with 70% sparsity are less promising. I believe further fine-tuning with sparsity is necessary

That sounds about right. For as you can see in [this blog](https://neuralmagic.com/blog/sparsegpt-remove-100-billion-parameters-for-free/), 60-70% sparsity is about the threshold where increased sparsity has significant effects on accuracy for open-ended LLM tasks.

**Comment by kylesayrs** [comment]
> Additionally, I’m curious whether the KV Cache can be quantized to the int8 format instead of fp8. This functionality is particularly important for me.

As far as I know, this functionality should be supported. I'm unfamiliar with whether kv-cache quantization is performed when inferencing using huggingface, you may need to use [vllm](https://github.com/vllm-project/vllm) for now.

@horheynm May be able to speak more about whether he's run this particular configuration and the current state of kv-cache quantization with huggingface.

**Comment by dsikka** [comment]
Hi @zjnyly - we do not support running quantized kv_caches in llmcompressor. You will have to load the model using vLLM to leverage the quantized cache

**Comment by jessiewiswjc** [comment]
@zjnyly @kylesayrs @dsikka Good jobs! Have you tested the accuracy of the model produced by sparsification? I am sad to find that my 14B model's mmlu score dropped a lot (from 72 to 60) after using sparsification.

**Comment by zjnyly** [comment]
> [@zjnyly](https://github.com/zjnyly) [@kylesayrs](https://github.com/kylesayrs) [@dsikka](https://github.com/dsikka) Good jobs! Have you tested the accuracy of the model produced by sparsification? I am sad to find that my 14B model's mmlu score dropped a lot (from 72 to 60) after using sparsification.

Hi, I make several experiments on Llama-3 8B model. I didn't keep the result, but I think when aligning to the same model size (50% sparsification W8A8 quantization vs W5A8 (sparse matrix need bitmask to record the position)), the latter always performs better (I use awq to quantize model).

**Comment by dsikka** [comment]
@jessiewiswjc @zjnyly 

The example for sparsification that we have provided is a basic example and will likely involve additional steps for finetuning. For additional details, please consider our Axolotl integration: https://developers.redhat.com/articles/2025/06/17/axolotl-meets-llm-compressor-fast-sparse-open which will allow you to leverage the Axolotl library for better finetuning capabilities
